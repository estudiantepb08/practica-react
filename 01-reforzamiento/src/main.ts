import './style.css';
import './bases/01-const-let';
import './bases/02-template-string';
import './bases/03-object-literal.ts';
import './bases/04-arrays.ts';
import './bases/05-funtions.ts';
import './bases/06-obj-destructuring.ts';
import './bases/07-array-destructuring.ts';
import { getHeroeByOwner } from './bases/08-imp-exp.ts'
import { Owner } from './data/heroes.data.ts';
import './bases/09-promises.ts';
import './bases/10-fetch-api.ts';
import './bases/11-async-await.ts';

document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
  <div>
    <h1>HOLA MUNDO</h1>
  </div>
`
console.log(getHeroeByOwner(Owner.Marvel));
console.log(getHeroeByOwner(Owner.DC));

// Importante: Si se importa algo que no se usa, da error. Por eso se importan los otros archivos

// Ejemplos de desestructuraci√≥n


