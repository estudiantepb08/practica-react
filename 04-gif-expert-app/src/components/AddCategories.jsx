import { useState } from "react";
import { PropTypes } from "prop-types";

export const AddCategories = ({onNewCategories})=>{

    const [inputValue, setInputValue] = useState('');

    const onInputChange = ({target}) =>{
        console.log(target.value); //Obtenemos el valor del campo por el target
        setInputValue(target.value);
    };

    const onSubmit = (event)=>{
        event.preventDefault();//con este metodo le decimos que no haga refresh al hacer enter en la tecla
        console.log(inputValue);
        /**validar el input con los valores vacios */
        if(inputValue.trim().length <= 1) return;
        //setCategories(categories => [inputValue, ...categories]); //obtenemos el valor del input y lo agregamos a la lista
        /**limpiamos el input */
        onNewCategories(inputValue.trim()); //Le asignamos el valor del input a la propiedad del props
        setInputValue(''); //Resetea el
        
    };

    return(
    <form onSubmit={(event) => onSubmit(event)} aria-label="form">
    <input type="text" placeholder="Buscar Gif"
    value={inputValue}
    onChange={onInputChange}/>
    </form>
    );
}

AddCategories.propTypes = {
    onNewCategories: PropTypes.func.isRequired,
}