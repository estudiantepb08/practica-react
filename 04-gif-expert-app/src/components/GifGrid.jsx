//Este componente va a manejar las categorias
/** Para desplegar la aplicaciÃ³n en produccion hay un paquete de node.js que se llama http-server
 * los archivos que se colocan para produccion son: 'asset index.html'
 * se compila el proyecto con el comando 'npm build'
 * para levantar el proyecto de forma local compilado es de la siguiente manera:
 * 1. Nos paramos en la carpeta dist de compilacion 
 * 2. Dentro de la carpeta ejecutamos el comando 'http-server -o'
 */

import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GitGrid = ({category})=>{  

    /* Custom Hook personalizado lo estamos inicializando con categori*/

    const {images, isLoading} = useFetchGifs(category);  

    return(<>
    
    <h3>{category}</h3>
    {/**Este es un if o and en JavaScript no usar operador ternario se carga cuando las imagenes no han sido resultas por la promesa con http: */
        isLoading && (<h2>Cargando....</h2>)
    }
    <div className="card-grid">{
    /* En este ejemplo estamos utilizando la desestructuracion del objeto de retorno images 
        images.map(({id, title}) =>(
            <li key={id}>{title}</li>
        ))
        */
       /** Otra forma de hacerlo mas optimizado 
        * una tecnica de esparcir las propiedades es con obtener el objeto completo y con el express podemos desestructurar todas sus propiedades
        * ejemplo: {...image} el componente GifItem puede utilizar las propiedades
       */

       images.map((image)=>(
        <GifItem key={image.id} { ...image }></GifItem>
       ))
        }        
    </div>
    </>);
}