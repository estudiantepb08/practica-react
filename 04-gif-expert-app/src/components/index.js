/**Es un archivo de barril para utilizar todos los componentes y hacer una solo importacion como lo hace react */

export * from './AddCategories';
export * from './GifGrid';
export * from './GifItem';