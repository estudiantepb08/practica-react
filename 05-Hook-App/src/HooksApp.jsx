export const HooksApp = ()=>{
    return(
        <h1>Hello Work</h1>
    );
}