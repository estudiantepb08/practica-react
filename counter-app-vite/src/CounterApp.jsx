import React from 'react';
import { useState } from 'react';

export const CounterApp = ({value})=>{
    /**
     * Creamos nuestro primer hook que es el estado el hook renderiza o dibuja el nuevo estado del componente
     * utiliza los argumentos el valor que va a dibujar y donde lo voy a setear con set()
     * siempre deben ser una constante
     * const [counter, setCounter] = useState(0);
     */
    const [counter, setCounter] = useState(value);

    const handleApp = () =>{setCounter(counter + 1)};
    const handleSubtract = () =>{setCounter(counter - 1)};
    const handleReset = () =>{setCounter(value)};
    /*
    se puede enviar solo la referencia a la funcion sin enviarle la cordenada completa
    tradicional
    <button onClick={(event) => handleApp(event)}> + 1</button>
    solo referencia en la funcion se pasan los argumentos de la misma forma
    <button onClick={handleApp}> + 1</button>
    */
   /**
    * Para realizar las pruebas unitarias necesitamos installar en VITE la libreria de JEST
    * Y en el package.json agregar la propiedad "test":"jest" "test": "jest --watchAll"
    * para ejecutar las pruebas es con el comando yarn test o npm run test 
    * otra manera de identificar un elemento HTML es con aria-label
    */

    return(
        <>
        <h1>Evento clic</h1>
        <h2>{ counter }</h2>
        <button onClick={ handleApp }> + 1</button>
        <button onClick={ handleSubtract }> - 1</button>
        <button aria-label='btn-reset' onClick={ handleReset }> Reset</button>
        </>
    );
}
