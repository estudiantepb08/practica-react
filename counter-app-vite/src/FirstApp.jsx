import React from 'react';

const newMessages = 'ALEX'; 
/*
esta constantes deben inr fuera de la funcion para que no se ha renderizado
No he permitodo los objetos json, solo es permitido serializandolo con la clase JSON.stringify
*/

const getResultado = (a,b) =>{
    return a + b;
};

export const FirstApp = () =>{
    return (
        <>
    <h2>First App</h2>
    <p>hola: { newMessages }</p>
    <p>Valor Numero: {getResultado(2,8)}</p>
    </>
    );
};