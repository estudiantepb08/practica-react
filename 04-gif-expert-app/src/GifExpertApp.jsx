import React from 'react';

import { useState } from "react";
import { AddCategories, GifGrid } from './components';

export const GifExpertApp = ()=>{

    const [categories, setCategories] = useState(['One Punch','Saludo']); 
    /** El valor inicial que pase en el useState lo obtengo con el segundo argumento es decir
     * categories seria mi getCategories, y setCategories me recibe el valor
     */

    const onAddCategory = (newCategories)=>{
        /**Validamos que la categoria no este repetidad en el arreglo */
        if(categories.includes(newCategories)) return;
        //categories.push('Que tal');
        setCategories([newCategories, ...categories]); // hacer copia de categoria y agregar el nuevo valor 'Forma express'
        //setCategories(cat => [...cat, 'valorant']); // con callback
//<button onClick={onAddCategory}>Agregar</button>
    };

    return(<>
    {/** Titulo */}
    <h1>GitExpertApp</h1>
    {/** Input */}
    <AddCategories onNewCategories={onAddCategory}></AddCategories>
    {/** Listado de Gif */}
    
    <ol>{
       /* Primera forma de poder listar la lista de categorias
        categories.map(category =>{
            return<li key={category}>{category}</li>
        })
        */
       /** Segunda forma de listar o recorrer la lista desestructurando con un componente */

       categories.map(category =>(
        <GifGrid key={category} category={category}></GifGrid> //Componente GifGrid categori para desacoplar el codigo
       ))
        }
        
        {/*<li>123</li>
        <li>XYZ</li>*/}
    </ol>
    {/** Gif Item */}
    </>);
}