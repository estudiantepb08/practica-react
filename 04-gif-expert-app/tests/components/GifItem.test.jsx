import { render, screen} from '@testing-library/react';
import { GifItem } from '../../src/components';

describe('Prueba del archivo <GifItem/>', ()=>{

    const pTitulo = 'Titulo P';
    const pUrl = 'http://localhost/';

    test('Prueba propiedades requeridas titulo urls', ()=>{       
                
        const { container } = render(<GifItem title={pTitulo} url={pUrl}></GifItem>);
        <GifItem title={pTitulo} />
        expect(container).toMatchSnapshot();      

    });

    test('Debe mostrar la imagen con el URL y el ALT indicado', ()=>{
        render(<GifItem title={ pTitulo } url={ pUrl }></GifItem>);
       // screen.debug();
       // expect(screen.getByRole('img').toBe()); Otra forma es haciendo la desestructuraciÃ³n
       const { src, alt } = screen.getByRole('img');
       expect(alt).toBe(alt);
       expect(src).toBe(pUrl);
    });

    test('Debe mostrar el titulo en el componente', ()=>{
        /** Se valida el texto */
        render(<GifItem title={pTitulo} url={pUrl}></GifItem>);
        expect(screen.getByText(pTitulo)).toBeTruthy();
    });
    
});