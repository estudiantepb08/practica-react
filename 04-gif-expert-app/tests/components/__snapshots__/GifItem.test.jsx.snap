// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del archivo <GifItem/> Prueba propiedades requeridas titulo urls 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Titulo P"
      src="http://localhost/"
    />
    <p>
      Titulo P
    </p>
  </div>
</div>
`;
